File: mcp_server_development/python/src/session13/schema_dump.py
"""tools/list を書き出す（Python 版）

  docker compose exec python python src/session13/schema_dump.py
  docker compose exec python python src/session13/schema_dump.py --write

クライアント側のスクリプトなので print() を使ってかまいません。
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import anyio

sys.path.insert(0, str(Path(__file__).resolve().parent))

from fixture_server import create_fixture_server  # noqa: E402
from harness import SNAPSHOT_DIR  # noqa: E402
from mcp import Client  # noqa: E402


async def main() -> None:
    async with Client(create_fixture_server()) as client:
        result = await client.list_tools()

    # by_alias=True で「電文と同じ camelCase」に戻す（属性は snake_case なので）
    payload = result.model_dump(mode="json", by_alias=True, exclude_none=True)
    serialized = json.dumps(payload, ensure_ascii=False, indent=2, sort_keys=True) + "\n"
    print(serialized)

    if "--write" in sys.argv:
        SNAPSHOT_DIR.mkdir(exist_ok=True)
        file = SNAPSHOT_DIR / "fixture-tools-list.json"
        file.write_text(serialized, encoding="utf-8")
        print(f"書き出しました: {file}")


if __name__ == "__main__":
    anyio.run(main)

File: mcp_server_development/node/src/session13/schema-dump.ts
/**
 * tools/list を人が読める形で書き出す
 *
 *   docker compose exec node npx tsx src/session13/schema-dump.ts
 *   docker compose exec node npx tsx src/session13/schema-dump.ts --write
 *
 * --write を付けるとベースラインを上書きします（スキーマを意図的に変えたときに使う）。
 * クライアント側のスクリプトなので stdout に書いてかまいません
 * （サーバープロセスでは stdout が JSON-RPC の通信路になるので禁止です）。
 */
import fs from "node:fs";
import path from "node:path";

import { createDocSearchServer } from "../mid01/create-server.js";
import { createFixtureServer } from "./fixture-server.js";
import { connectInMemory } from "./harness.js";

const SNAPSHOT_DIR = path.resolve("src/session13/__snapshots__");
const shouldWrite = process.argv.includes("--write");

const targets = [
  { name: "mid01-tools-list", server: () => createDocSearchServer({ docsRoot: path.resolve("src/mid01/docs") }) },
  { name: "fixture-tools-list", server: () => createFixtureServer() },
];

fs.mkdirSync(SNAPSHOT_DIR, { recursive: true });

for (const target of targets) {
  const client = await connectInMemory(target.server(), "schema-dump");
  const { tools } = await client.listTools();
  await client.close();

  const serialized = `${JSON.stringify(tools, null, 2)}\n`;
  console.log(`===== ${target.name} =====`);
  console.log(serialized);

  if (shouldWrite) {
    const file = path.join(SNAPSHOT_DIR, `${target.name}.json`);
    fs.writeFileSync(file, serialized, "utf8");
    console.log(`書き出しました: ${file}`);
  }
}
